### PR TITLE
[GENX][JointMatrix]: Add the genx.matrix.mad operation to the GENX dialect

### PR DESCRIPTION
**Pull request by etiotto**
The genx.MatrixMad operation can be used to perform a multiply-add on Joint Matrices.

### CODE REVIEW

**Comment by etiotto** [comment]
This PR is based on PR #18

**Comment by etiotto** [comment]
@whitneywhtsang @pengtu this is ready for review, thanks.

File: mlir/include/mlir/Dialect/LLVMIR/GENXOps.td [hunk 1]
```diff
@@ -410,4 +410,27 @@ def GENX_MatrixLoadOp : GENX_Op<"matrix.load">,
...
+    the result. Matrix $a is a M x K matrix, $b is a K x N matrix and $c is a M x N matrix. 
+    The scope operand indicates the synchronization scope for operations on the matrix.
+  }];
+
+  string llvmBuilder = [{
+    $res = createMatrixMad(builder, $a, $b, $c, $scope);
+  }];
+
+  let assemblyFormat = [{
+    $scope operands attr-dict`:` type($a) `,` type($b) `->` type($c)
```
Comment (whitneywhtsang) [inline mlir/include/mlir/Dialect/LLVMIR/GENXOps.td]: Why do you prefer to print the types like this? shouldn't the operator takes 3 arguments and return 1, even if `type($c)` and `type($res)` are the same?

File: mlir/lib/Dialect/LLVMIR/IR/GENXOps.cpp [hunk 2]
```diff
@@ -30,3 +30,32 @@ LogicalResult GENX::MatrixLoadOp::verify() {
...
+  if (CType != resType)
+    return this->emitOpError("result and 3rd operand must have the same type");
+
+  if (AType.getNumRows() != resType.getNumRows() ||
+      AType.getNumColumns() != BType.getNumRows() ||
+      BType.getNumColumns() != resType.getNumColumns())
+    return this->emitOpError("matrix sizes must match");
+
+  if (AType.getElementType() != BType.getElementType() ||
+      resType.getElementType() != CType.getElementType())
```
Comment (whitneywhtsang) [inline mlir/lib/Dialect/LLVMIR/IR/GENXOps.cpp]: we already checked `CType == resType`, let's check
```suggestion
      AType.getElementType() != CType.getElementType())
```

File: mlir/test/Dialect/LLVMIR/genx-invalid.mlir [hunk 3]
```diff
@@ -13,3 +13,29 @@ func.func @joint_matrix_load(%ptr : !llvm.ptr<i32>, %stride : index) {
...
+func.func @joint_matrix_mad(%a : !genx.jointmatrix<8x16xi32, RowMajor>, %b : !genx.jointmatrix<8x8xi32, RowMajor>, %c : !genx.jointmatrix<8x8xi32, RowMajor>) {
+  // expected-error @+1 {{'genx.matrix.mad' op matrix sizes must match}}
+  %r = genx.matrix.mad <Subgroup> %a, %b, %c : !genx.jointmatrix<8x16xi32, RowMajor>, !genx.jointmatrix<8x8xi32, RowMajor> -> !genx.jointmatrix<8x8xi32, RowMajor>
+  llvm.return    
+}
+
+// -----
+
+func.func @joint_matrix_mad(%a : !genx.jointmatrix<8x16xf32, RowMajor>, %b : !genx.jointmatrix<16x8xi32, RowMajor>, %c : !genx.jointmatrix<8x8xi32, RowMajor>) {
+  // expected-error @+1 {{matrix element types must match}}
```
Comment (whitneywhtsang) [inline mlir/test/Dialect/LLVMIR/genx-invalid.mlir]: do you know why `'genx.matrix.mad' op` is not printed in this error?

File: mlir/test/Dialect/LLVMIR/genx.mlir [hunk 4]
```diff
@@ -119,3 +119,10 @@ func.func @genx.matrix.load(%ptr : !llvm.ptr<vector<4xi32>>, %stride : index) {
   %0 = genx.matrix.load <Subgroup> <RowMajor> %ptr, %stride {memory_access = #genx.memory_access<Volatile>} : (!llvm.ptr<vector<4xi32>>, index) -> !genx.jointmatrix<8x16xi32, RowMajor>
   llvm.return  
 }
+
+func.func @genx.matrix.mad(%a : !genx.jointmatrix<8x32xi8, RowMajor>, %b : !genx.jointmatrix<32x8xi8, ColumnMajor>, %c : !genx.jointmatrix<8x8xi32, RowMajor>) {
+  // CHECK-LABEL: genx.matrix.mad
+  // CHECK: %0 = genx.matrix.mad <Subgroup> %arg0, %arg1, %arg2 : !genx.jointmatrix<8x32xi8, RowMajor>, !genx.jointmatrix<32x8xi8, ColumnMajor> -> !genx.jointmatrix<8x8xi32, RowMajor>
+  %0 = genx.matrix.mad <Subgroup> %a, %b, %c : !genx.jointmatrix<8x32xi8, RowMajor>, !genx.jointmatrix<32x8xi8, ColumnMajor> -> !genx.jointmatrix<8x8xi32, RowMajor>
```
Comment (whitneywhtsang) [inline mlir/test/Dialect/LLVMIR/genx.mlir]: is that allowed to add matrix of `i8` with matrix of `i32`?

File: mlir/include/mlir/Dialect/LLVMIR/GENXOps.td (hunk 1 above)
Comment (etiotto) [reply mlir/include/mlir/Dialect/LLVMIR/GENXOps.td]: The res type and the type of matrix "C" are the same, that is the reason the type of C isn't printed.  May be easier to just print the type of all the operands and the result type. And prob. clearer, will do that.

File: mlir/test/Dialect/LLVMIR/genx-invalid.mlir (hunk 3 above)
Comment (etiotto) [reply mlir/test/Dialect/LLVMIR/genx-invalid.mlir]: fixed.

File: mlir/test/Dialect/LLVMIR/genx.mlir (hunk 4 above)
Comment (etiotto) [reply mlir/test/Dialect/LLVMIR/genx.mlir]: Yes the combination is valid (https://github.com/intel/llvm/blob/54cd00d559a89931e4bcc5cf0dbf5938e2767c52/sycl/doc/extensions/experimental/sycl_ext_matrix/sycl_ext_oneapi_matrix.asciidoc#intel-xmx-supported-combinations)

File: mlir/lib/Dialect/LLVMIR/IR/GENXOps.cpp (hunk 2 above)
Comment (etiotto) [reply mlir/lib/Dialect/LLVMIR/IR/GENXOps.cpp]: Here I think we need to check that the type combination is valid according to allowed combinations in https://github.com/intel/llvm/blob/54cd00d559a89931e4bcc5cf0dbf5938e2767c52/sycl/doc/extensions/experimental/sycl_ext_matrix/sycl_ext_oneapi_matrix.asciidoc#intel-xmx-supported-combinations

File: mlir/include/mlir/Dialect/LLVMIR/GENXOps.td (hunk 1 above)
Comment (etiotto) [reply mlir/include/mlir/Dialect/LLVMIR/GENXOps.td]: done

File: mlir/lib/Dialect/LLVMIR/IR/GENXOps.cpp (hunk 2 above)
Comment (etiotto) [reply mlir/lib/Dialect/LLVMIR/IR/GENXOps.cpp]: done and more verification tests have been added

File: mlir/include/mlir/Dialect/LLVMIR/GENXOps.td [hunk 5]
```diff
@@ -410,4 +410,27 @@ def GENX_MatrixLoadOp : GENX_Op<"matrix.load">,
...
+    the result. Matrix $a is a M x K matrix, $b is a K x N matrix and $c is a M x N matrix. 
+    The scope operand indicates the synchronization scope for operations on the matrix.
+  }];
+
+  string llvmBuilder = [{
+    $res = createMatrixMad(builder, $a, $b, $c, $scope);
+  }];
+
+  let assemblyFormat = [{
+    $scope operands attr-dict`:` type(operands) `->` type($res)
```
Comment (whitneywhtsang) [inline mlir/include/mlir/Dialect/LLVMIR/GENXOps.td:456]: Is below possible?
```suggestion
    $scope operands attr-dict`:` functional-type(operands, results)
```

**Review by whitneywhtsang — APPROVED** [review]

**Review by pengtu — APPROVED** [review]

File: mlir/include/mlir/Dialect/LLVMIR/GENXOps.td (hunk 5 above)
Comment (etiotto) [reply mlir/include/mlir/Dialect/LLVMIR/GENXOps.td:456]: No because I do not want to have parenthesis around the operands.

**Pull request merged**